helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventThreadPoolFactory.java

```diff
@@ -0,0 +1,98 @@
+package org.apache.helix.manager.zk;
+
+
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+
+// returns a thread pool object given a HelixManager identifier
+public class CallbackEventThreadPoolFactory {
+  private static final int CALLBACK_EVENT_THREAD_POOL_SIZE = 10;
+  private static final int CALLBACK_EVENT_THREAD_POOL_TTL_MINUTE = 3;
+  static private final ReadWriteLock _lock = new ReentrantReadWriteLock();
+  static private Map<Integer, ThreadPoolExecutor> _managerToCallBackThreadPoolMap = new HashMap();
+  static private Map<Integer, AtomicInteger> _callBackEventProcessorCountPerThreadPool =
+      new HashMap();
+
+  public static ThreadPoolExecutor getOrCreateThreadPool(int hash) {
+    // should not use general lock for read
+    _lock.readLock().lock();
+    ThreadPoolExecutor result = null;
+    if (_managerToCallBackThreadPoolMap.containsKey(hash)) {
+      result = _managerToCallBackThreadPoolMap.get(hash);
+      _callBackEventProcessorCountPerThreadPool.get(hash).incrementAndGet();
+    }
+    _lock.readLock().unlock();
+    if (result == null) {
+      result = getOrCreateThreadPoolHelper(hash);
+    }
+    return result;
+  }
+
+  private static ThreadPoolExecutor getOrCreateThreadPoolHelper(int hash) {
+    ThreadPoolExecutor result;
+    _lock.writeLock().lock();
```

Comment:
Better to put the unlock call in a final block to ensure safety.

Reply:
TFTR. 
Please correct if I am wrong. I think putting this unlock call to final block wont down grade the lock to a read lock. This need to happen before the 
`result = _managerToCallBackThreadPoolMap.get(hash);`

Reply:
Not the expected unlock, that one is fine.
What I meant is that, in general, if an exception is thrown unexpectedly, then try-catch-finally will help to ensure the lock is released instead of being held forever. An alternative way is to use some autoclosable lock design. In your current code, one example is that if the new threadpoolexecutor fails, then the writelock won't be released.

Reply:
TFTR. Good point. Updated.